### ISSUE TITLE
[Question]: Accessing Photos Saved in the Clipboard Using the Plugin

### ISSUE THREAD

**Issue by SameerJS6** [CoreyHayward/PowerToys-Run-ClipboardManager#17]
Is there a way to access photos saved in the clipboard through this plugin?

For example, while we can access photos from the Windows clipboard, we cannot access those images through this plugin.

Window's Clipboard: 
![image](https://github.com/user-attachments/assets/5f50e6e6-e2f2-44f5-bd6b-ca795656ea8b)

Clipboard Manger Plugin.
![image](https://github.com/user-attachments/assets/6d1f0c90-9813-46af-bbdf-c4c2f17113b0)

**Comment by CoreyHayward** [comment]
Currently no.

Since PowerToys Run is text based the power that the Clipboard Manager plugin enables is the ability to search the clipboard history which isn't really possible with images.

The other main issue with images being that there would be no way to preview them since PowerToys Run (afaik) doesn't support images in the results main content.

However, it may be possible to display them instead of the icon on the left with some generic text.

For the moment it's not planned but I may take a look into this at some point and see if it's even possible.

**Comment by SameerJS6** [comment]
@CoreyHayward Thank you for clarifying that!

I have a small question: how difficult is it to develop a plugin for PowerToys? I don’t have much experience with native Windows development, but I believe we’d need to use `C#` to build plugins, right?

Additionally, is it possible for a **TypeScript** developer to build a plugin that can take any documentation URL, fetch all the main topics or sub-routes, and make them accessible in a way similar to the TailwindCSS plugin for PowerToys, but for general-purpose use?

Tailwind Plugin SS:
![image](https://github.com/user-attachments/assets/47744a56-f85a-44d7-b1bf-bd1a8d057975)

**Comment by CoreyHayward** [comment]
@SameerJS6 Developing plugins for PowerToys isn't too difficult, with the hardest thing being that the documentation is not particularly extensive however there are some helpful guides and quite a few examples of different plugins which mostly cover all of the functionality you might need to write one.

You are correct that plugins do currently have to be written in C# but if your coming from TypeScript you should be able to mostly read C# as the syntax is somewhat similar and there is a helpful guide by Microsoft here which will help you get started with C# -> https://learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/tips-for-javascript-developers

Your plugin idea sounds super cool but also quite complex as each site may have a different mechanism for sub topics/routes and you would need a way to parse those automatically which would be quite tough if its not standardised in any way.

If you do want to give developing a plugin a go here are some links which will help you along the way:
https://github.com/hlaueriksson/awesome-powertoys-run-plugins
https://conductofcode.io/post/creating-custom-powertoys-run-plugins/
https://github.com/microsoft/PowerToys/blob/main/doc/devdocs/modules/launcher/new-plugin-checklist.md

And if you need any specific help along the way feel free to reach out and I will try to be of assistance!

**Comment by SameerJS6** [comment]
@CoreyHayward Can you share your Twitter or Discord handle (If you don't mind)? I've started with C# in Visual Studio with the basics and will be starting with the blog (the one you shared; Thank you!). Communicating via Twitter or Discord messages would be better for both of us than using GitHub issues. 

Also, what things would you recommend I should be aware of before building a plugin that interacts with a third-party REST API? I researched a bit and found out about: `Try/Catch` (similar to TS/JS), Error handling.

In the JS ecosystem, we have many `HTTP` clients/packages to handle various aspects of making asynchronous requests (Such as [`axios`](https://axios-http.com/docs/intro) or complete package like [`Tanstack Query`](https://tanstack.com/query/latest). Is there anything like that in C#?

**Comment by SameerJS6** [comment]
Yo!! @CoreyHayward, I've built my first PowerToys Plugin!!

It's called SVGL For PowerToys.
[Website](https://svgl.sameerjs.com/) | [GitHub Repo](https://github.com/SameerJS6/powertoys-svgl)

Thank You!! Brother, my helping in the beginning for providing really really helpful resources, specifically that Creating custom PowerToys Plugin by @hlaueriksson

However, it's not the plugin that I had discussed with you, mainly due to the limitation of PowerToys (No Async support for Query method, etc). Its a plugin that allows you to search, browse, and copy tech companies' logos in SVG format. It was something that I used to use a lot in my job, so I thought, why not just build it?

![Image](https://github.com/user-attachments/assets/bebb857b-e38c-44b4-bb4e-465e582cb224)

**Comment by CoreyHayward** [comment]
Hey @SameerJS6,

I am so sorry I didn't see your last comment, must have missed it in my notifications so really sorry for not getting back to you.

Your plug-in looks really cool and super useful - I'm definitely going to download and use it!

The website you created for it is also super clean! Did you use a template for that or made from scratch?

Great work, and I'm glad I was able to help in some way!

Corey

**Comment by SameerJS6** [comment]
@CoreyHayward No need to apologize! To be honest, that question seems pretty dumb now that I think about it. 😂

I'll be waiting to hear about your experience using it! I did my best to make it as fast as possible by leveraging caching and minimizing async operations. The first time you initialize the plugin, it might take a few milliseconds, but after that, it loads instantly.

Regarding the website: I think you might not remember, but I'm a web developer first. I designed and developed the entire website myself. It's built using Next.js, Tailwind CSS, and Framer Motion.

![Image](https://github.com/user-attachments/assets/ca1b3a95-0f77-4a50-8ac2-90ae3ec8bf45)
